src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs

```diff
@@ -108,72 +138,74 @@ void IRunningDocumentTableEventListener.OnRenameDocument(string newMoniker, stri
             {
             }
 
-            public static async Task<OpenFileTracker> CreateAsync(VisualStudioWorkspaceImpl workspace, IAsyncServiceProvider asyncServiceProvider)
+            public static async Task<OpenFileTracker> CreateAsync(
+                VisualStudioWorkspaceImpl workspace,
+                IThreadingContext threadingContext,
+                IAsyncServiceProvider asyncServiceProvider,
+                IAsynchronousOperationListenerProvider listenerProvider)
             {
                 var runningDocumentTable = (IVsRunningDocumentTable?)await asyncServiceProvider.GetServiceAsync(typeof(SVsRunningDocumentTable)).ConfigureAwait(true);
                 Assumes.Present(runningDocumentTable);
 
                 var componentModel = (IComponentModel?)await asyncServiceProvider.GetServiceAsync(typeof(SComponentModel)).ConfigureAwait(true);
                 Assumes.Present(componentModel);
 
-                return new OpenFileTracker(workspace, runningDocumentTable, componentModel);
+                return new OpenFileTracker(workspace, threadingContext, runningDocumentTable, componentModel, listenerProvider);
             }
 
             private void TryOpeningDocumentsForMoniker(string moniker, ITextBuffer textBuffer, IVsHierarchy? hierarchy)
             {
                 _foregroundAffinitization.AssertIsForeground();
 
-                _workspace.ApplyChangeToWorkspace(w =>
+                _workspaceApplicationQueue.AddWork(async () =>
                 {
                     var documentIds = _workspace.CurrentSolution.GetDocumentIdsWithFilePath(moniker);
```

Comment:
There's a potential bug here now since this CurrentSolution is not being read under the lock, and the document could get removed between here and when the lock is acquired. I see a few options:

1. Just put the whole thing under the ApplyChangeToWorkspaceMaybeAsync call -- that means yes we're doing a UI thread switch while holding the lock which isn't ideal, but it's at least all async.
2. Try to refactor GetActiveContextProjectIdAndWatchHierarchiesAsync to not need the candidate ProjectIds. It looks reasonably doable I think; instead of it returning the ProjectId it might just return the hierarchy and maybe the value from VSHPROPID_ActiveIntellisenseProjectContext, and then we look that up once we've acquired the lock.

Reply:
> that means yes we're doing a UI thread switch while holding the lock which isn't ideal, but it's at least all async.

That definitely worries me... and makes me feel we *might* have some chance of deadlock.  Though i suppose it's hard to get a lock inversion here as nothing can block on the BG work here...

Reply:
> Try to refactor GetActiveContextProjectIdAndWatchHierarchiesAsync to not need the candidate ProjectIds.

Trying this first.

Reply:
Ok.  that worked.  not pretty.  but it worked.